server/automute_test.go

```diff
@@ -0,0 +1,451 @@
+	})
+
+	t.Run("should return true when a user is connected and we skip checking their primary platform", func(t *testing.T) {
+		p := newAutomuteTestPlugin(t)
+
+		mockUserConnected(p, user.Id)
+
+		// This would cause it to return false if we didn't skip the check
+		setPrimaryPlatform(p, user.Id, PreferenceValuePlatformMM)
+
+		result, err := p.shouldEnableAutomuteForUser(user.Id, false, true)
+
+		assert.Equal(t, true, result)
+		assert.NoError(t, err)
+	})
+
+	t.Run("should return false when a user is not connected and we're skipping checking their primary platform", func(t *testing.T) {
+		p := newAutomuteTestPlugin(t)
+
+		mockUserNotConnected(p, user.Id)
+
+		// This would cause it to return false if we didn't skip the check
+		setPrimaryPlatform(p, user.Id, PreferenceValuePlatformMM)
+
+		result, err := p.shouldEnableAutomuteForUser(user.Id, false, true)
+
+		assert.Equal(t, false, result)
+		assert.NoError(t, err)
+	})
+
+	t.Run("should return true when a user has Teams as their primary platform and we're skipping checking if they're connected", func(t *testing.T) {
+		p := newAutomuteTestPlugin(t)
+
+		// This assumes that calling isUserConnected causes the mock to panic
+
+		setPrimaryPlatform(p, user.Id, PreferenceValuePlatformMSTeams)
+
+		result, err := p.shouldEnableAutomuteForUser(user.Id, true, false)
+
+		assert.Equal(t, true, result)
+		assert.NoError(t, err)
+	})
+
+	t.Run("should return true when a user has Teams as their primary platform and we're skipping checking if they're connected", func(t *testing.T) {
+		p := newAutomuteTestPlugin(t)
+
+		// This assumes that calling isUserConnected causes the mock to panic
+
+		setPrimaryPlatform(p, user.Id, PreferenceValuePlatformMM)
+
+		result, err := p.shouldEnableAutomuteForUser(user.Id, true, false)
+
+		assert.Equal(t, false, result)
+		assert.NoError(t, err)
+	})
+
+	t.Run("should return true when we skip both checks", func(t *testing.T) {
+		p := newAutomuteTestPlugin(t)
+
+		// This assumes that calling isUserConnected causes the mock to panic
+
+		// This would cause it to return false if we didn't skip the check
+		setPrimaryPlatform(p, user.Id, PreferenceValuePlatformMM)
+
+		result, err := p.shouldEnableAutomuteForUser(user.Id, true, false)
+
+		assert.Equal(t, false, result)
+		assert.NoError(t, err)
+	})
+}
+
+func TestSetAutomuteEnabledForUser(t *testing.T) {
+	p := newAutomuteTestPlugin(t)
+
+	user := &model.User{Id: model.NewId()}
+
+	channel, _ := p.API.CreateChannel(&model.Channel{Id: model.NewId(), Type: model.ChannelTypeOpen})
+	p.API.AddUserToChannel(channel.Id, user.Id, "")
+	directChannel, _ := p.API.GetDirectChannel(user.Id, model.NewId())
+
+	mockLinkedChannel(p, channel)
+
+	t.Run("initial conditions", func(t *testing.T) {
+		assertChannelNotAutomuted(t, p, channel.Id, user.Id)
+		assertChannelNotAutomuted(t, p, directChannel.Id, user.Id)
+	})
+
+	t.Run("should do nothing when false is passed and automuting has never been enabled", func(t *testing.T) {
```

Comment:
Wonder if this set of tests could be refactor to do it in a table testing way (i.e. define the test cases, and then run them in a for loop).

Reply:
Table testing wouldn't work in this case since it's actually relying on the state from the previous test for each step of this. The reason it uses `t.Run` is to group the assertions together so that it's more clear what each set of assertions does. I could instead write it like
```go
	p := newAutomuteTestPlugin(t)

	user := &model.User{Id: model.NewId()}

	channel, _ := p.API.CreateChannel(&model.Channel{Id: model.NewId(), Type: model.ChannelTypeOpen})
	p.API.AddUserToChannel(channel.Id, user.Id, "")
	directChannel, _ := p.API.GetDirectChannel(user.Id, model.NewId())

	mockLinkedChannel(p, channel)

	// Confirm initial conditions
		assertChannelNotAutomuted(t, p, channel.Id, user.Id)
		assertChannelNotAutomuted(t, p, directChannel.Id, user.Id)
	})

	// Confirm that this should do nothing when false is passed and automuting has never been enabled
	result, err = p.setAutomuteEnabledForUser(user.Id, false)

	assert.Equal(t, false, result)
	assert.NoError(t, err)

	assertChannelNotAutomuted(t, p, channel.Id, user.Id)
	assertChannelNotAutomuted(t, p, directChannel.Id, user.Id)
```
but I like brackets and indentation to help differentiate those sets.

Plus it makes the output look more interesting 😛